### DISCUSSION TITLE
Resolved: Add a `test-case` crate

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#3568] [Ideas]
Previously we have discussed the idea of somehow providing test cases for various structs so our test code doesn't include so many arbitrary hex strings (and other could use it too). But we never ironed out the issues around it.

New Idea: What about a crate that lives in `rust-bitcoin/test-case` and includes just hex strings as consts e.g.,

```rust
mod chain {
    mod block {
        pub const BLOCK_700_001 = "<hex>"; // Can we use `include_str!` in const context?
        pub const BLOCK_700_002 = "<hex>";
        pub const BLOCK_700_003 = "<hex>";

        pub const TX_1 = "hex for txout vout 1 from block 700,002";
        pub const TX_2 = "hex for txout vout 2 from block 700,002";
    }
}

mod dummy {
    pub const BLOCK = "hex representing some made up dummy values";
    pub const TRANSACTION = "hex representing some made up dummy values";
    // Same for PRIVATE_KEY PUBLIC_KEY etc
}
```

`rust-bitcoin` can then have a dev dependency on `test-case` and we can merge whatever const people want into it.

**Comment by apoelstra** [comment]
I think we probably want `test-case` to depend on `bitcoin` and to have a bunch of unit tests in it, to the extent possible.

Pretty much all Bitcoin data is pretty transparent, so if we find stuff that we can't do without private access to our types, that's a strong hint that our API is incomplete. (For example consider #3567.)

**Reply by tcharding** [reply]
We can't create `Block` if we make the fields private, that is how I got to this idea.

**Reply by tcharding** [reply]
and also `bitcoin` can't use the crate if it depends on `bitcoin`.

**Reply by apoelstra** [reply]
> We can't create `Block` if we make the fields private, that is how I got to this idea.

If `test-case` can't create `Block`s then neither can anybody else :). We need to have constructors.

**Reply by tcharding** [reply]
Oh I should have clarified, in test code there are a bunch of _invalid_ blocks, if we introduce invariants then it becomes hard to make those - though now I write this that makes no sense deserializing hex will get invariants enforced too.

**Reply by apoelstra** [reply]
We can attempt to deserialize the invalid blocks, and they should fail with an appropriate error message. But there's no point in trying to test what happens if you manage to construct an invalid block, since you shouldn't be able to.

**Comment by apoelstra** [comment]
Relevant BIP to block parsing https://github.com/bitcoin/bips/blob/master/bip-0034.mediawiki

**Comment by tcharding** [comment]
This discussion is dead in the water, marked as "Resolved: " in title, I'm not sure how else to mark this as "closed".

**Reply by apoelstra** [reply]
I'd be okay leaving it open. But yes, I think this has been mostly subsumed by us implementing `arbitrary` for a ton more types.
